### DISCUSSION TITLE
Interface design discussion

### DISCUSSION THREAD

**Discussion by mcastorina** [vlang/v#15189] [Ideas]
I've noticed a few warts with interfaces so I wanted to start a discussion to nail down their goals. I'll start with how I expect interfaces to behave.

1. Interfaces are a contractual binding for functions and methods (i.e. "I don't care what specific type it is, as long as it has these methods / attributes"
2. You should never have to cast a concrete type to an interface - the type either implements it or not
3. Interfaces should be embeddable
4. Interfaces can have default implementations (i.e. "If you define x, you get y for free")
5. Default implementations may be overwritten by a concrete type's implementation
    a. I'm not 100% certain this should be allowed

I think in its current state interfaces are mostly there, but the warts I have noticed:
* having to cast to an interface for one reason or another
* interfaces seem to be types of their own

**Comment by larpon** [comment]
I like it at first glance but also feel like 4 and 5 need to be really thought through
